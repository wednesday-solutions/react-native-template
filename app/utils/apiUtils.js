import { create } from 'apisauce';
import mapKeysDeep from 'map-keys-deep';
import { camelCase, snakeCase } from 'lodash';
import { Config } from '@app/config/index';

export const apiClients = {
  configApi: null,
  default: null
};
export const getApiClient = (type = 'configApi') => apiClients[type];
export const generateApiClient = (type = 'configApi') => {
  switch (type) {
    case 'configApi':
      apiClients[type] = createApiClientWithTransForm(Config.API_URL);
      return apiClients[type];
    default:
      apiClients.default = createApiClientWithTransForm(Config.API_URL);
      return apiClients.default;
  }
};

export const createApiClientWithTransForm = baseURL => {
  const api = create({
    baseURL,
    headers: { 'Content-Type': 'application/json' }
  });
  api.addResponseTransform(response => {
    const { ok, data } = response;
    if (ok && data) {
      response.data = mapKeysDeep(data, keys => camelCase(keys));
    }
    return response;
  });

  api.addRequestTransform(request => {
    const { data } = request;
    if (data) {
      request.data = mapKeysDeep(data, keys => snakeCase(keys));
    }
    return request;
  });
  return api;
};
