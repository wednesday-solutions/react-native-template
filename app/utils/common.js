export const conditionalOperatorFunction = (condition, val1, val2) =>
  condition ? val1 : val2;
