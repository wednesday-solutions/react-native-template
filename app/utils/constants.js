export const LoadingStates = {
  LOADING: 'loading',
  HAS_ERROR: 'hasError',
  HAS_VALUE: 'hasValue'
};
