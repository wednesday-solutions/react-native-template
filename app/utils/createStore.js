import { applyMiddleware, compose, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { persistReducer, persistStore } from 'redux-persist';
import immutableTransform from 'redux-persist-transform-immutable';
/**
 * This import defaults to localStorage for web and AsyncStorage for react-native.
 *
 * Keep in mind this storage *is not secure*. Do not use it to store sensitive information
 * (like API tokens, private and sensitive data, etc.).
 *
 * If you need to store sensitive information, use redux-persist-sensitive-storage.
 * @see https://github.com/CodingZeal/redux-persist-sensitive-storage
 */

const persistConfig = {
  key: 'root',
  transforms: [immutableTransform()],
  storage: AsyncStorage,
  /**
   * Blacklist state that we do not need/want to persist
   */
  blacklist: [
    // 'auth',
  ]
};

export default (rootReducer, rootSaga) => {
  // connect sagas to redux store
  const sagaMiddleware = createSagaMiddleware();
  const middleware = [sagaMiddleware];
  const enhancers = [applyMiddleware(...middleware)];

  // Redux persist
  const persistedReducer = persistReducer(persistConfig, rootReducer);

  const store = createStore(persistedReducer, compose(...enhancers));
  const persistor = persistStore(store);

  // Kick off the root saga
  sagaMiddleware.run(rootSaga);

  return { store, persistor };
};
