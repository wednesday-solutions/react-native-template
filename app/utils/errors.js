export const Errors = {
  USER_FETCH_ERROR: "'There was an error while fetching user information.'"
};
