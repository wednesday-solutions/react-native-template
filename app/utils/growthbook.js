import { GrowthBook } from '@growthbook/growthbook';
import { GROWTH_BOOK_API_HOST, GROWTH_BOOK_CLIENT_KEY } from '@env';

let growthBookClient;

/**
 * creates an instance of the growthBook client
 * @returns growthBook client
 */
export const createGrowthBookClient = userEmail => {
  const growthBook = new GrowthBook({
    apiHost: GROWTH_BOOK_API_HOST,
    clientKey: GROWTH_BOOK_CLIENT_KEY,
    enableDevMode: true,
    subscribeToChanges: true
  });

  growthBook.setAttributes({
    email: userEmail
  });

  return growthBook;
};

/**
 *  Function to get the GrowthBook client instance
 * @returns {GrowthBook} growthBook instance
 */
export function getGrowthBookClient(email) {
  if (growthBookClient) {
    return growthBookClient;
  }
  return createGrowthBookClient(email);
}

/**
 * Function to get  growthBook feature
 * @param {String} email
 * @returns {Applicant}  growthBook feature value
 */
export async function getGrowthBookFeaturesData(name, email) {
  try {
    const growthBook = getGrowthBookClient(email);
    await growthBook.loadFeatures();
    return growthBook.getFeatureValue(name);
  } catch (error) {
    return new Error(error);
  }
}

/**
 * Function to get growthBook feature
 * @param {String} name
 * @returns {Boolean} growthBook feature status
 */
export async function getGrowthBookFeatureFlag(name, email) {
  try {
    const growthBook = getGrowthBookClient(email);
    await growthBook.loadFeatures();
    return growthBook.isOn(name);
  } catch (error) {
    return new Error(error);
  }
}
