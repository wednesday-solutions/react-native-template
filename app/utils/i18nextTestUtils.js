import i18next from 'i18next';
import { initReactI18next } from 'react-i18next';
/* eslint-disable fp/no-mutating-assign */
/**
 * i18n.js
 *
 * This will setup the i18n language files and locale data for your app.
 *
 *   IMPORTANT: This file is used by the internal build
 *   script `extract-intl`, and must use CommonJS module syntax
 *   You CANNOT use import/export in this file.
 */
// // const addLocaleData = require('react-intl').addLocaleData; //eslint-disable-line

// // const enLocaleData = require('react-intl/locale-data/en');

const enTranslationMessages = require('@app/translations/en.json');

const languageDetector = {
  type: 'languageDetector',
  async: true,
  detect: cb => cb('en'),
  init: () => {},
  cacheUserLanguage: () => {}
};

i18next
  .use(languageDetector)
  .use(initReactI18next)
  .init({
    fallbackLng: 'en',
    debug: true,
    resources: {
      en: {
        translation: enTranslationMessages // log this object to ensure it's correctly loaded
      }
    },
    interpolation: {
      escapeValue: true // not needed for react!!
    }
  });
