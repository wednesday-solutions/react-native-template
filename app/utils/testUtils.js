import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import createStore from 'app/rootReducer';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@atoms/LanguageProvider';

export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});
export const renderWithIntl = (children, renderFunction = render) =>
  renderFunction(
    <IntlProvider
      locale={DEFAULT_LOCALE}
      messages={translationMessages[DEFAULT_LOCALE]}
    >
      {children}
    </IntlProvider>
  );

export const renderProvider = children => {
  const { store } = createStore();
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        {children}
      </ConnectedLanguageProvider>
    </Provider>
  );
};

export const timeout = ms => new Promise(resolve => setTimeout(resolve, ms));
