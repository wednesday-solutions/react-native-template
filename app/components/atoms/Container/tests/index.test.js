/**
 *
 * Tests for Container
 *
 */
import React from 'react';
import { renderWithI18next } from 'app/utils/testUtils';
import Container from '../index';

describe('<Container />', () => {
  it('should render and match the snapshot', () => {
    const baseElement = renderWithI18next(<Container />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 container', () => {
    const { getAllByTestId } = renderWithI18next(<Container />);
    expect(getAllByTestId('container').length).toBe(1);
  });
});
