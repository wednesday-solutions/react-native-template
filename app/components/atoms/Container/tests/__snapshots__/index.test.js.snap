// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> should render and match the snapshot 1`] = `
<View
  style={
    {
      "display": "flex",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
  testID="container"
/>
`;
